tools/circle_plus_gen/lib/train_info.py

```diff
@@ -0,0 +1,91 @@
+import json
+
+from schema import circle_traininfo_generated as ctr_gen
+'''
+Wrappers of the auto generated classes in circle_traininfo_generated
+
+Wrapper classes provides additional interfaces(e.g. initializer) to the auto generated classes.
+'''
+
+# Optimizers
+
+
+class SGD(ctr_gen.SGDOptionsT):
+    name = ['sgd', 'stocasticgradientdescent']
+
+
+class Adam(ctr_gen.AdamOptionsT):
+    name = ['adam']
+
+
+# Loss
+
+
+class SparseCategoricalCrossEntropy(ctr_gen.SparseCategoricalCrossentropyOptionsT):
+    name = [
+        'sparse categorical crossentropy', 'sparsecategoricalcrossentropy', 'sparsecce'
+    ]
+
+
+class CategoricalCrossEntropy(ctr_gen.CategoricalCrossentropyOptionsT):
+    name = ['categorical crossentropy', 'categoricalcrossentropy', 'cce']
+
+
+class MeanSqauaredError(ctr_gen.MeanSquaredErrorOptionsT):
+    name = ['mean squared error', 'mse']
+
+
+# TrainInfo
+
+
+class TrainInfo(ctr_gen.ModelTrainingT):
+    TRAINING_FILE_IDENTIFIER = b"CTR0"
+
+    def __init__(self):
+        super().__init__()
+
+    @classmethod
+    def from_buff(cls, buff):
+        '''Create TrainInfo from buffer(byte array)'''
+        tinfo = super().InitFromPackedBuf(bytearray(buff))
+        tinfo.__class__ = TrainInfo
```

Comment:
ditto.

Reply:
(Optional) Seems that current class design makes it hard to extend and hard to know what it is for. This is implementation detail but IMHO, you may just introduce a dump class. Dump class know the dumping extension and its format.

```python
TransInfoDump tid = TransInfoDump(TRAINING_FILE_IDENTIFIER)

# load model
# m: ModelTrainingT

tid.dump(m)
```

Reply:
> (Optional) Seems that current class design makes it hard to extend and hard to know what it is for. 

I hope that this becomes clear with the next PR.

I thought that `TrainInfo` is a wrapper class of `ModelTrainingT`.
It is to give additional method for easy use of `ModelTrainingT`: 

The additional method is likes..  
  * constructors for easy to create ModelTrainingT from various formats,  
  * converting method from the instance to the packed(serialized) buffer 

```python 
class TrainInfo(ModelTrainingT):
	def from_buff(): 
		# create traininfo from packed buffer
	def from_json(): 
		# create traininfo from json file 
		# will be added next PR 
	def dump_as_json(): 
		# dump to the json format 
	def to_buffer():
		# convert traininfo to packed buffer 
		# will be added next PR 
```

>  IMHO, you may just introduce a dump class 

Since I have a plan to add some additional methods, introducing a dump class might be a little hard.